### ISSUE TITLE
Feature Request Adaptive streaming

### ISSUE THREAD

**Issue by alturismo** [segator/proxylive#21]
Hi, so, after some testings and so on (high cpu usage) may a request about hls streaming

to get some benfit from hls it would be nice to implement adaptive streaming, so running as sample

360 and 480 transcode and build the m3u8 with both streams ... so HLS could play its benefit ;)

**Comment by segator** [comment]
Hls it was an experiment its not stable, probsbly needs to be rebuilded from scratch, Im afraid i dont have time for this :( feel free to make PR and I happily will review it

**Comment by alturismo** [comment]
cant make any pr on java at all

as note about abr with vaapi, sample here running smooth and stable

`ffmpeg -hwaccel vaapi -vaapi_device /dev/dri/renderD128 -hwaccel_output_format vaapi \
  -nostdin -hide_banner -loglevel quiet \
  -i "$URL" -y \
  -vf format=vaapi,hwupload,deinterlace_vaapi,scale_vaapi=854:-1 \
  -c:v h264_vaapi -c:a aac -ar 48000 -ac 2 -map 0:0 -map 0:1 -map 0:0 -map 0:1 -map 0:0 -map 0:1\
  -c:v:0 h264_vaapi -b:v:0 400k \
  -c:v:1 h264_vaapi -b:v:1 800k \
  -c:v:2 h264_vaapi -b:v:2 1200k \
  -c:a:0 aac -ar 48000 -ac 2 -b:a:0 128k \
  -var_stream_map "v:0,a:0 v:1,a:1 v:2,a:2" \
  -master_pl_name master.m3u8 \
  -f hls -hls_time 6 -hls_list_size 10 -hls_flags delete_segments \
  -hls_segment_filename "v%v/fileSequence%d.ts" \
  v%v/prog_index.m3u8`

but this requires then to serve the master.m3u8 only ... 

so i can of course get the ffmpeg line into your application.aml, but i dont get the webserver part to serve the master.m3u8 instead ...

**Comment by teminvictus** [comment]
@alturismo , would you share your php solution?

**Comment by alturismo** [comment]
hmm, when i find some time i can do so, currently theres nothing documentated yet, builded it into another docker of mine locally here, i ll look into it and drop a comment here when i find some time.

**Comment by segator** [comment]
I'm considering to build from scratch the HLS/Dash adapter any tought?

**Comment by alturismo** [comment]
@segator yes ;) the majot goal in my case was ABR (adaptive streaming rates) so it really makes use of HLS/DASH

in my usecase i have it currently as config part like this, input m3u, 3 bitrates which are then used by connection quality, a fixed resolution die i didnt wanted to add also dynamic resolutions depending on bitrate ;) specially depending on players this can cause issues.

`<?php
// streamlink, sample https://yourdomain.com/?streamtype=hls&streamlink=123456789abc

// fetch args from link local or remote
if ($_GET) {
    $streamtype = $_GET['streamtype'];
    $streamlink = $_GET['streamlink'];
} else {
    $streamtype = $argv[1];
    $streamlink = $argv[2];
}

// setup inputstream
$inputline = "http://localhost:34400/stream/$streamlink";	// format inputline
//
$bitratelow = "400k";	// lowwest bitrate
$bitratemed = "750k";	// medium  bitrate
$bitratehigh = "1100k";	// highest bitrate
//
$resolution = "854";	// resolution width, Aspect ratio is kept

// end config here`
...

which ends up in some bigger ffmpeg strings ...

![image](https://user-images.githubusercontent.com/8406490/118348154-a8406f80-b548-11eb-984e-75b67927b9ee.png)

and this only makes sense with hw accel ffmpeg transcoding, in software its just too much to handle

in case i can help out a bit ...
